{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b246cda3-a8ed-4b09-9343-b91c86455c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mplpatches\n",
    "import numpy as np\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd66b755-17ad-4d8a-a6a6-90c351b5183b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, '/private/groups/brookslab/gabai/tools/seqUtils/src/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c9704b79-944f-4692-b7b5-fb512c461279",
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqUtil import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "03ae7023-52cd-44fe-9853-fe6a15a73056",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetchSize(genome):\n",
    "    '''Input a genome.fa file, get the chromsome bounderies.'''\n",
    "    \n",
    "    genomeSize = {}\n",
    "    sequence = ''\n",
    "    with open(genome, 'r') as refFh:\n",
    "        for line in refFh:\n",
    "            line = line.strip()\n",
    "            if '>' in line:\n",
    "                if sequence:\n",
    "                    genomeSize[chrom] = len(sequence)\n",
    "                    sequence = ''\n",
    "                    chrom = line.split('>')[1]\n",
    "                else:\n",
    "                    chrom = line.split('>')[1]\n",
    "            else:\n",
    "                sequence += str(line)\n",
    "        if sequence:\n",
    "            genomeSize[chrom] = len(sequence)\n",
    "    return genomeSize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "2843052d-ec08-413c-970c-c2947ebbca2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrVI_bam = '/private/groups/brookslab/gabai/projects/Add-seq/data/chrom/mapping/chrVI_chrom.sorted.bam'\n",
    "genome = '/private/groups/brookslab/gabai/projects/Add-seq/data/ref/sacCer3.fa'\n",
    "chrVI_reads = getAlignedReads(chrVI_bam, region = 'chrVI', genome = genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "403d3d53-335b-4da8-a9dc-dfb7a0d53416",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrVI_evt = '/private/groups/brookslab/gabai/projects/Add-seq/data/chrom/eventalign/chrVI.eventalign.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "187a6dda-0040-45e4-9785-3d90533d4a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "parseEventAlign(eventAlign = chrVI_evt, outfile= '/data/scratch/gabai/addseq_data/eventalign/chrVI.eventalign.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2816fb30-bae2-4e4f-85d8-49a21c5bed41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parseEventAlign(eventAlign = '', outfile = '', print_sequence = False, header = True):\n",
    "    '''\n",
    "    This function reads nanopolish eventalign file, aggregates signals and the number of \n",
    "    signals correspinding to one base movement for read in readname list.\n",
    "    \n",
    "    input:\n",
    "        eventAlign: nanopolish eventalign output file.\n",
    "        readname: A list containing readnames.\n",
    "        chr_region: chromosome number that region of interest falls in.\n",
    "    optional:\n",
    "        print_sequence: if True, kmer sequence will be included in outfile.\n",
    "        n_rname: number of readnames can be skipped in the readname list (default: 0).\n",
    "                 Searching all the readnames from the eventalign file takes longer time.\n",
    "    output: \n",
    "        outfile: siganlAlign.tsv with format: readname\\tchrom\\teventStart(reference)\\tsigList\\tsigLenLsit\n",
    "\n",
    "    E.g.    read1  ACGTGGCTGA\n",
    "            events ACGTG\n",
    "                    CGTGG\n",
    "                     GTGGC\n",
    "                      TGGCT\n",
    "                       GGCTG\n",
    "                        GCTGA\n",
    "            sigLen  23\n",
    "                     45\n",
    "                      61\n",
    "                       78\n",
    "                        101\n",
    "    '''\n",
    "    \n",
    "    if outfile:\n",
    "        outf = open(outfile, 'w')\n",
    "    \n",
    "    tag = ''\n",
    "    read = ''\n",
    "    sequence = ''\n",
    "    c = 0\n",
    "    \n",
    "    with open(eventAlign, 'r') as inFile:\n",
    "        if header:\n",
    "            header = inFile.readline()\n",
    "        for line in inFile:\n",
    "            line = line.strip().split('\\t')\n",
    "            thisread = line[3]\n",
    "            thischrom = line[0]\n",
    "            c+=1\n",
    "            if c%10000000 == 0:\n",
    "                print(c/1000000, ' M lines have passed.')\n",
    "\n",
    "            if thisread != read:\n",
    "\n",
    "                # parsed read exist\n",
    "                if sequence:\n",
    "                    # Set variables back to initial state\n",
    "                    if print_sequence:\n",
    "                        out = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(read, chrom, eventStart, sequence, ','.join(str(i) for i in sigList), ','.join(str(i) for i in sigLenList))\n",
    "                    else:\n",
    "                        out = \"{}\\t{}\\t{}\\t{}\\t{}\\n\".format(read, chrom, eventStart, ','.join(str(i) for i in sigList), ','.join(str(i) for i in sigLenList))\n",
    "                    if outfile:\n",
    "                        outf.write(out)\n",
    "                    read = ''\n",
    "                    sequence = ''\n",
    "                    sigList = []\n",
    "                    sigLenList = []\n",
    "\n",
    "                # very first read\n",
    "                read = thisread\n",
    "                chrom = thischrom\n",
    "                eventStart = line[1]\n",
    "                start = line[1]\n",
    "                kmer = line[2]\n",
    "\n",
    "                # signals are stored in column 13/15 and are separated my comma\n",
    "                sigList = [float(i) for i in line[-1].split(',')]\n",
    "                sigLen = len(sigList)\n",
    "                sigLenList = [sigLen]\n",
    "                sequence = kmer\n",
    "            \n",
    "            # next kmer within the same read\n",
    "            else:\n",
    "                signals = [float(i) for i in line[-1].split(',')]\n",
    "                # or signalList += signals\n",
    "                sigList.extend(signals)\n",
    "                # signalLength records the number of signals for one base movement\n",
    "                sigLen += len(signals)\n",
    "\n",
    "                # If different kmer\n",
    "                if (line[1], line[2]) != (start, kmer):\n",
    "                    deletion = int(line[1]) - int(start) - 1\n",
    "                    # id there is a deletion in eventalign file\n",
    "                    if deletion > 0:\n",
    "                        sequence += deletion*'D'\n",
    "                        for i in range(deletion):\n",
    "                            sigLenList.append(sigLenList[-1])\n",
    "                    start = line[1]\n",
    "                    kmer = line[2]\n",
    "                    sequence += kmer[-1]\n",
    "                    sigLenList.append(sigLen)\n",
    "                # If same kmer\n",
    "                else:\n",
    "                    # Update the number of signals matched to previous kmer\n",
    "                    sigLenList[-1]=sigLen\n",
    "        if sequence:\n",
    "            if print_sequence:\n",
    "                out = \"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(read, chrom, eventStart, sequence, ','.join(str(i) for i in sigList), ','.join(str(i) for i in sigLenList))\n",
    "            else:\n",
    "                out = \"{}\\t{}\\t{}\\t{}\\t{}\\n\".format(read, chrom, eventStart, ','.join(str(i) for i in sigList), ','.join(str(i) for i in sigLenList))\n",
    "            if outfile:\n",
    "                outf.write(out)\n",
    "    outf.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
