{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b246cda3-a8ed-4b09-9343-b91c86455c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mplpatches\n",
    "import numpy as np\n",
    "import sys\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "bd66b755-17ad-4d8a-a6a6-90c351b5183b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, '/private/groups/brookslab/gabai/tools/seqUtils/src/')\n",
    "from seqUtil import *\n",
    "from nanoUtil import *\n",
    "from bamUtil import *\n",
    "from nntUtil import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "acbaff27-40d2-489a-8581-0b370a7f4001",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom, qStart, qEnd\n",
    "step = 40\n",
    "signalWindow = 400\n",
    "kmerWindow = 75\n",
    "bins = np.arange(qStart, qEnd, step)\n",
    "models = {\n",
    "            'resnet1D':resnet1D\n",
    "            }\n",
    "mymodel = models['resnet1D']\n",
    "model = mymodel\n",
    "\n",
    "myweight =  '/private/groups/brookslab/gabai/tools/seqUtils/src/nanopore_classification/best_models/addseq_resnet1d.pt'\n",
    "weight = myweight"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae78e85-b41d-4f6b-847d-d2120ed22015",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input files\n",
    "bam = '/private/groups/brookslab/gabai/projects/Add-seq/data/chrom/mapping/chrom.sorted.bam'\n",
    "genome = '/private/groups/brookslab/gabai/projects/Add-seq/data/ref/sacCer3.fa'\n",
    "chrII_evt = '/data/scratch/gabai/addseq_data/eventalign/chrII.eventalign.txt'\n",
    "\n",
    "# output files\n",
    "chrII_sig= '/prfivate/groups/brookslab/gabai/projects/Add-seq/data/chrom/eventalign/rand50_chrII.sigAlign.tsv'\n",
    "\n",
    "chrII_alignment, chrom, qStart, qEnd = getAlignedReads(bam, region = 'chrII', genome = genome)\n",
    "parseEventAlign(eventAlign = chrII_alignment, alignment = chrII_alignment, outfile =chrII_sig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "7c55d3f0-b4c4-4bfb-828a-440da3206805",
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.arange(qStart, qEnd, step)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "18c93064-e333-49c8-b917-7b6aba76e5a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class findNemo:\n",
    "\n",
    "    '''\n",
    "    class findNemo: predict small molecule modifications from nanopore long-read sequencing data.\n",
    "    '''\n",
    "    \n",
    "    def __init__(self, region, bam, genome, outpath, prefix, eventalign = '', sigalign = ''):\n",
    "        \n",
    "        '''\n",
    "        self:\n",
    "            self.alignment: dict, stores reference an reads mapped to region.\n",
    "            self.chrom: str, chromosome name.\n",
    "            self.qStart: int, query start position.\n",
    "            self.qEnd: int, query end position\n",
    "            self.outpath: str, path to store the output files.\n",
    "        Input:\n",
    "            region: genomic coordinates to perform modification predictions. Format: 'chrI:2000-5000' or 'chrI'.\n",
    "            bam: sorted, indexed, and binarized alignment file.\n",
    "            genome: reference genome fasta file.\n",
    "        Output:\n",
    "            outpath: path to store the output files.\n",
    "            prefix: prefix attached to the output file names.\n",
    "            eventalign: nanopolish eventalign file.\n",
    "            sigalign: sigalign file if sigAlign file already exist. If not, must provide eventalign to generate sigAlign file.\n",
    "        '''\n",
    "\n",
    "        # Fetch reads aligned to the region\n",
    "        self.alignment, self.chrom, self.qStart, self.qEnd = getAlignedReads(bam, region = region, genome = genome)\n",
    "        self.outpath = outpath\n",
    "\n",
    "        # Index reads to avoid storing the long readnames. \n",
    "        self.reads = {r:i for r,i in zip(list(self.alignment.keys())[1:], range(len(self.alignment)-1))}\n",
    "        # Store the id index match into a file.\n",
    "        readFh = open(outpath + prefix + region + '_readID.tsv', 'w')\n",
    "        for k,v in self.reads: readFh.write('{read}\\t{index}\\n'.format(k,v))\n",
    "        readFh.close()\n",
    "        print(len(self.reads), \" reads mapped to \", region)\n",
    "\n",
    "        if sigalign:\n",
    "            self.sigalign = sigalign\n",
    "        elif eventalign:\n",
    "            self.sigalign = outpath + prefix + str(region) + '_sig.tsv'\n",
    "            parseEventAlign(eventAlign = eventalign, reads = self.reads, outfile = self.sigalign)\n",
    "        else:\n",
    "            print('Error: None of sigalign or eventalign files are provided!')\n",
    "\n",
    "    def doWork(work):\n",
    "    \n",
    "        (readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight) = work\n",
    "        \n",
    "        scores = runNNT(readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight)\n",
    "        \n",
    "        return scores\n",
    "    \n",
    "    def modPredict(model, weight, threads, step=40, kmerWindow=75, signalWindow=400, load = 1000):\n",
    "        \n",
    "        bins = np.arange(self.qStart, self.End, kmerWindow)\n",
    "        \n",
    "        device = \"cuda\" if torch.cuda.is_available() else \"mps\" if torch.backends.mps.is_available() else \"cpu\"\n",
    "        print('Device type: ', device)\n",
    "        \n",
    "        # Use the specified threads number or maximum available CPU cores\n",
    "        num_processes = min(threads, multiprocessing.cpu_count())\n",
    "        pool = multiprocessing.Pool(processes=num_processes)\n",
    "\n",
    "        # Total work to be done are stored in a list\n",
    "        works = [(readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight) \n",
    "                  for readID, aStart, aEnd, strand, sigList, sigLenList in parseSigAlign(self.sigalign, self.alignment)]\n",
    "        \n",
    "        for x in range(0, len(works), load):\n",
    "            # split total work by load\n",
    "            works_per_load = works[x:x+load]\n",
    "            \n",
    "            predOut = outpath + prefix + str(region) + '_' + str(x) + '_prediction.tsv'\n",
    "\n",
    "            # Use the pool.map() function to process reads in parallel\n",
    "            outs = pool.map(self.doWork, works_per_load)\n",
    "\n",
    "            # Close the pool to release resources\n",
    "            pool.close()\n",
    "            pool.join()\n",
    "\n",
    "            # Write the results from current work load\n",
    "            predOutFh = open(predOut, 'w')\n",
    "            for r in range(len(outs)):\n",
    "                out = outs[r]\n",
    "                readID = works_per_load[r][0]\n",
    "                bin_start = next(iter(out))\n",
    "                predOutFh.write('{readID}\\t{bin_start}\\t{scores}'.format(readID = readID, bin_start = bin_start, scores = ','.join(map(str, out.values()))))\n",
    "            predOutFh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "c0e51581-8d0c-4962-b301-1c992f93e398",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeBedGraph(bedGraphHeader, binScores, chrom, outfile, binCounts = '', step = 40, normalize = False):\n",
    "    outFh = open(outfile, 'w')\n",
    "    \n",
    "    for k,v in bedGraphHeader.items():\n",
    "        if v:\n",
    "            line = k + '=' + v + ' '\n",
    "            outFh.write(line)\n",
    "    outFh.write('\\n')\n",
    "    \n",
    "    for chrStart in binScores.keys():\n",
    "        chrEnd = chrStart + step\n",
    "        if normalize:\n",
    "            score = \"%.3f\" % (binScores[chrStart]/binCounts[chrStart])\n",
    "        else:\n",
    "            score = \"%.3f\" % (binScores[chrStart])\n",
    "        \n",
    "        line = '{chr}\\t{start}\\t{end}\\t{score}\\n'.format(chr = chrom, start = chrStart,  end = chrEnd, score = score)\n",
    "        outFh.write(line)\n",
    "    outFh.close()\n",
    "\n",
    "bedGraphHeader = {'track type':'bedGraph', \n",
    "                  'name':'test', \n",
    "                  'description':'test1read',\n",
    "                  'visibility':'', \n",
    "                  'color':'r', \n",
    "                  'altColor':'r', \n",
    "                  'priority':'', \n",
    "                  'autoScale':'off', \n",
    "                  'alwaysZero':'off', \n",
    "                  'gridDefault':'off', \n",
    "                  'maxHeightPixels':'default', \n",
    "                  'graphType':'bar',\n",
    "                  'viewLimits':'upper',\n",
    "                  'yLineMark':'',\n",
    "                  'yLineOnOff':'on',\n",
    "                  'windowingFunction':'mean',\n",
    "                  'smoothingWindow':'on'\n",
    "                 }\n",
    "\n",
    "writeBedGraph(bedGraphHeader, binScores=out[0], chrom = chrom, outfile = '../data/chrom/modPredict/231110_test.bedgraph')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "0a8f846c-0ac4-48ec-9a9c-077e891d2119",
   "metadata": {},
   "outputs": [],
   "source": [
    "worker_inputs = [(readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight) \n",
    "                  for readID, aStart, aEnd, strand, sigList, sigLenList in parseSigAlign(chrII_sig, chrII_alignment)]\n",
    "\n",
    "(readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight) = worker_inputs[0]\n",
    "binScores = runNNT(readID, bins, step, aStart, aEnd, sigList, sigLenList, kmerWindow, signalWindow, device, model, weight)\n",
    "out = [binScores]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
